src/main/clojure/jarl/builtins/jwt.clj

```diff
@@ -0,0 +1,308 @@
+    (try
+      (base64/url-decode-bytes s)
+      (catch ExceptionInfo e
+        (if (errors/builtin-ex? e)
+          (throw (errors/builtin-ex "JWT %s had invalid encoding: %s" part (strip-ex-type (ex-message e))))
+          (throw e))))))
+
+(defn jwk-key-resolver ^VerificationKeyResolver [^String s]
+  (reify VerificationKeyResolver
+    (resolveKey [_ _ _]
+      (try
+        (.getKey (JsonWebKey$Factory/newJwk s))
+        (catch Throwable _
+          (throw (errors/builtin-ex "failed to parse a JWK key (set)")))))))
+
+(defn strict-jwks-key-resolver ^VerificationKeyResolver [^String s]
+  (-> (JsonWebKeySet. s)
+      (.getJsonWebKeys)
+      (JwksVerificationKeyResolver.)
+      (doto (.setDisambiguateWithVerifySignature true))))
+
+(defn valid? [^JsonWebSignature jws ^JsonWebKey key]
+  (and (= (.getAlgorithmHeaderValue jws) (.getAlgorithm key))
+       (.verifySignature (doto jws (.setKey (.getKey key))))))
+
+(defn- try-strict-resolve [^String s jws nesting-context]
+  (try
+    (.resolveKey (strict-jwks-key-resolver s) jws nesting-context)
+    (catch UnresolvableKeyException _ false)))
+
+(defn jwks-key-resolver [^String s]
+  (reify VerificationKeyResolver
+    (resolveKey [_ jws nesting-context]
+      (if-let [key (try-strict-resolve s jws nesting-context)]
+        (.getKey ^JsonWebKey key)
+        (.getKey ^JsonWebKey (first (filter #(valid? jws %)
+                                            (-> s (JsonWebKeySet.) (.getJsonWebKeys)))))))))
+
+(defn- es-alg? [alg]
+  (contains? #{:es256 :es384 :es512} alg))
+
+(defn- hs-alg? [alg]
+  (contains? #{:hs256 :hs384 :hs512} alg))
+
+(defn pem-key-resolver [^String s alg]
+  (let [key-spec (-> s
+                     (str/replace "-----BEGIN PUBLIC KEY-----" "")
+                     (str/replace "-----END PUBLIC KEY-----" "")
+                     (str/replace (System/lineSeparator) "")
+                     (base64/decode-bytes)
+                     (X509EncodedKeySpec.))
+        public-key (.generatePublic (KeyFactory/getInstance (if (es-alg? alg) "EC" "RSA")) key-spec)]
+    (reify VerificationKeyResolver
+      (resolveKey [_ _ _]
+        (if (es-alg? alg)
+          (cast ECPublicKey public-key)
+          (cast RSAPublicKey public-key))))))
+
+(defn x509-key-resolver [^String s]
+  (if-not (or (str/ends-with? s "-----END CERTIFICATE-----")
+              (str/ends-with? s "-----END CERTIFICATE-----\n"))
+    (throw (errors/builtin-ex "extra data after a PEM certificate block"))
+    (try
+      (let [der (strip-pem-wrapper s)]
+        (-> (X509Util.)
+            (.fromBase64Der der)
+            (List/of)
+            (X509VerificationKeyResolver.)
+            (doto (.setTryAllOnNoThumbHeader true))))
+      (catch Throwable _
+        (throw (errors/builtin-ex "failed to parse a PEM certificate"))))))
+
+(defn key-resolver-name [^String certificate]
+  (if (str/starts-with?     certificate "{")
+    (if (str/includes?      certificate "keys") "jwks" "jwk")
```

Comment:
Doesn't this run the risk of false positives where a JWK contains `keys` somewhere within it?

Reply:
Hehe, that's true I suppose :) I'll change that to be `\"keys\"`.

Reply:
Done.